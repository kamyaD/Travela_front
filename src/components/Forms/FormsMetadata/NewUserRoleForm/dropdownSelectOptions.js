const dropdownSelectOptions = {
  roleName: [
    'Super Administrator',
    'Travel Administrator',
    'Travel Team Member',
    'Requester',
    'Manager'
  ]
};

export default dropdownSelectOptions;
