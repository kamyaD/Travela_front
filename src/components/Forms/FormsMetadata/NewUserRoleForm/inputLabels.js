const inputLabels = {
  email: {
    label: 'Andela Email Address',
  },
  roleName: {
    label: 'Role'
  },
  center: {
    label: 'Center'
  },
};

export default inputLabels;
