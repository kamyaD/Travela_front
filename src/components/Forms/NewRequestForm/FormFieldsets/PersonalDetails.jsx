import React, { Component } from 'react';
import PropTypes from 'prop-types';
import InputRenderer from '../../FormsAPI';
import * as formMetadata from '../../FormsMetadata/NewRequestFormMetadata';
import expand from '../../../../images/expand_more_24px.svg';
import Checkbox from '../../../CheckBox/index';

class PersonalDetailsFieldset extends Component {
  state = {
    disableInputs: ''
  };
  handleDisableInputs = value => {
    const newState = value === 'clicked' ? 'disable-details' : '';
    this.setState({
      disableInputs: newState
    });
  };
  renderfields = collapse => {
    const { disableInputs } = this.state;
    const { values, value, managers, occupations } = this.props;
    const managerChoices = managers.map(manager => manager.fullName);
    const occupationsNames = occupations.map(
      occupation => occupation.occupationName
    );
    const { renderInput } = this.inputRenderer;
    const disabled = disableInputs;
    return (
      <div>
        {!collapse ? (
          <div>
            <div className={`input-group ${disabled}`}>
              <div className="spaces">{renderInput('name', 'text')}</div>
              <div className="spaces">
                {renderInput('gender', 'button-toggler')}
              </div>
              {renderInput('department', 'dropdown-select', { size: value, className: 'request_dropdown'})}
            </div>
            <div className={`input-group ${disabled}`}>
              <div className="spaces">
                {renderInput('role', 'filter-dropdown-select', {
                  choices: occupationsNames,
                  size: value,
                  className: 'request_dropdown',
                  id: 'your-role'
                })}
              </div>
              <div className="spaces">
                {renderInput('manager', 'filter-dropdown-select', {
                  choices: managerChoices,
                  size: value,
                  className: 'request_dropdown',
                  id: 'your-manager'
                })}
              </div>
            </div>
            <div className="input-group">
              <Checkbox
                values={values}
                handleDisableInputs={this.handleDisableInputs}
              />
            </div>
          </div>
        ) : null}
      </div>
    );
  };

  render() {
    const { collapsible, collapse, title, position, line, values } = this.props;
    this.inputRenderer = new InputRenderer(formMetadata);
    const { renderInput } = this.inputRenderer;

    const disabledFields =
      values.state === 'clicked' ? 'disable-details' : null;
    return (
      <fieldset className={`personal-details ${disabledFields}`}>
        <legend
          style={{
            width: '100%',
            borderBottom: line,
            fontFamily: 'DIN Pro Medium',
            fontSize: '18px',
            paddingTop: '12px'
          }}>
          Personal Details
          <span className="required-field">* Required Field</span>
          <span
            className="hide-details"
            onClick={collapsible}
            onKeyPress={this.handleKeyDown}
            role="button"
            tabIndex={0}
            style={{ outline: 'none' }}>
            <img src={expand} alt="clicked" className="expand" style={{ transform: position }} />
            {!title ? 'Hide Details' : title}
          </span>
        </legend>
        {this.renderfields(collapse)}
      </fieldset>
    );
  }
}

const managers = PropTypes.array;
const collapsible = PropTypes.func;
const collapse = PropTypes.bool;
const title = PropTypes.string;
const position = PropTypes.string;
const line = PropTypes.string;
const values = PropTypes.object;
const occupations = PropTypes.array;

PersonalDetailsFieldset.propTypes = {
  managers: managers.isRequired,
  collapsible: collapsible.isRequired,
  collapse: collapse.isRequired,
  title: title.isRequired,
  position: position.isRequired,
  line: position.isRequired,
  values: values,
  value: PropTypes.string,
  occupations
};

PersonalDetailsFieldset.defaultProps = {
  values: {},
  value: '',
  occupations: []
};

export default PersonalDetailsFieldset;
