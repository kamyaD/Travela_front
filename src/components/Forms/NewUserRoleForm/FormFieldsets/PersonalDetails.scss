.add_button{
    border: 1px solid #3359DB;
    border-radius: 1px;
    background-color: #3359DB;
    height: 36px;
    width: auto;
    border: 1px solid #D4D4D4;
    border-radius: 1px;
    color: white;
    font-family: "DIN Pro";
    font-size: 14px;
    font-weight: 500;
    line-height: 14px;
    margin: 28px 1px 0pc 352px;
    padding: 9px 21px;
    cursor: pointer;
    outline: none !important;
}

.size_in{
    padding-top: 14px;
    display: flex;
    margin-bottom: 17px;
}


.back{
    background-color: white !important;
}

.remove_department{
    font-family: "DIN Pro";
    font-style: normal;
    font-size: 20px !important;
    font-weight: 900 !important;
    color: #FF5359;
    display: inline-block;
    outline: none;
    width: 25px;
    cursor: pointer;
}

.disable_button {
    background-color: rgba(51, 89, 219, 0.3);
    border: none;
    color: #FFFFFF;
    cursor: not-allowed;
    height: 36px;
    width: auto;
    font-family: "DIN Pro";
    font-size: 14px;
    font-weight: 500;
    line-height: 14px;
    margin: 28px 1px 0pc 352px;
    padding: 9px 21px;
}

.add_dept {
    position: fixed;
    width: 65%;
}



