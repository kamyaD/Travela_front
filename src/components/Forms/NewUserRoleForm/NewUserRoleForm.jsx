import React, { PureComponent } from 'react';
import toast from 'toastr';
import { PropTypes } from 'prop-types';
import { FormContext } from '../FormsAPI';
import PersonalDetailsFieldset from './FormFieldsets/PersonalDetails';
import SubmitArea from '../NewRequestForm/FormFieldsets/SubmitArea';

class NewUserRoleForm extends PureComponent {
  constructor(props) {
    super(props);
    const { role, userDetail, roleId } = this.props;
    let defaultValues = {
      email: (userDetail || {}).email,
      roleName: role,
      department: '',
      departments: (userDetail.budgetCheckerDepartments || []).map(dept => dept.name),
      roleId,
    };
    this.defaultState = {
      values: defaultValues,
      errors: {},
      hasBlankFields: true
    };
    this.state = { ...this.defaultState };
  }

  componentDidMount() {
    const { getAllUsersEmail } = this.props;
    getAllUsersEmail();
  }

  componentWillUnmount() {
    const { getRoleData } = this.props;
    this.handleCancel();
    getRoleData();
  }

  handleSubmit = event => {
    event.preventDefault();
    const { handleUpdateRole, updateBudgetChecker, myTitle } = this.props;
    const { values } = this.state;
    const dataValue = {
      email: values.email,
      roleName: values.roleName,
      departments: values.departments
    };
    if (this.validate()) {
      myTitle === 'Add User' ?
        handleUpdateRole(dataValue):
        updateBudgetChecker({email: dataValue.email, departments: values.departments});
    }
  };


  addDepartment = () => {
    const { values } = this.state;
    const testPattern = new RegExp(`^${values.department.trim()}$`, 'i');
    if (values.departments.find(value => testPattern.test(value))) {
      return toast.error('You have added that department already');
    }

    this.setState({
      values: {
        ...values,
        departments: values.departments.concat([values.department.trim()]),
        department: ''
      }
    },  this.validate
    );
  }

  removeDepartment = (i) => {
    const { values } = this.state;
    const array = values.departments;
    array.splice(i, 1);
    this.setState({
      values: {
        ...values,
        departments: array
      }
    }, this.validate
    );
  }

  handleCancel = () => {
    const { closeModal } = this.props;
    this.setState({ ...this.defaultState });
    closeModal();
  };

  validate = field => {
    let { values, errors } = this.state;
    const { role } =  this.props;
    [errors, values] = [{ ...errors }, { ...values }];
    let hasBlankFields = false;
    delete values.department;
    if (!values[field]) {
      errors[field] = 'This field is required';
      delete errors.department;
    } else {
      errors[field] = '';
    }

    hasBlankFields = Object.keys(values).some(key => !values[key]);
    this.setState(prevState => {
      return { ...prevState, errors, hasBlankFields };
    });

    if(role === 'Budget Checker'  && values.departments.length < 1){
      this.setState({
        hasBlankFields:true
      });
    }
    return !hasBlankFields;
  };

  render() {
    const { values, errors, hasBlankFields } = this.state;
    const { updatingRole, role, centers, myTitle, allMails } = this.props;
    return (
      <FormContext targetForm={this} values={values} errors={errors} validatorName="validate">
        <form onSubmit={this.handleSubmit} className="new-request">
          <PersonalDetailsFieldset
            values={values}
            roleName={role}
            centers={centers}
            addDepartment={this.addDepartment}
            hasBlankFields={hasBlankFields}
            removeDepartment={this.removeDepartment}
            myTitle={myTitle}
            allMails={allMails}
          />
          <SubmitArea
            onCancel={this.handleCancel}
            loading={updatingRole}
            hasBlankFields={hasBlankFields}
            send="Submit"
          />
        </form>
      </FormContext>
    );
  }
}

NewUserRoleForm.propTypes = {
  handleUpdateRole: PropTypes.func.isRequired,
  updatingRole: PropTypes.bool,
  getRoleData: PropTypes.func.isRequired,
  centers: PropTypes.array.isRequired,
  myTitle: PropTypes.string.isRequired,
  updateBudgetChecker: PropTypes.func,
  role: PropTypes.string.isRequired,
  userDetail:  PropTypes.object,
  roleId: PropTypes.string.isRequired,
  getAllUsersEmail: PropTypes.func,
  allMails: PropTypes.array.isRequired,
  closeModal: PropTypes.func.isRequired
};

NewUserRoleForm.defaultProps = {
  updatingRole: false,
  updateBudgetChecker: ()=> {},
  getAllUsersEmail: ()=> {},
  userDetail: {},
};

export default NewUserRoleForm;
