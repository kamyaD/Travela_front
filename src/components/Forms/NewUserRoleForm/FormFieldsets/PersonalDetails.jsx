import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import InputRenderer from '../../FormsAPI';
import * as formMetadata from '../../FormsMetadata/NewUserRoleForm';
import './PersonalDetails.scss';

class PersonalDetailsFiedset extends Component {
  render() {
    const { roleName, values, allMails, addItem, removeItem, myTitle, departments, centers } = this.props;
    const emails = allMails.map(email => email.text);
    const allDepartments = departments.map(dept => dept.text);
    const allCenters = centers.map(center => center.location);
    const dropDown = roleName === 'Budget Checker' ? allDepartments : allCenters;
    const disabled = myTitle.split(' ')[0] === 'Edit';
    formMetadata.dropdownSelectOptions.email = emails;
    formMetadata.dropdownSelectOptions.item = dropDown;
    this.inputRenderer = new InputRenderer(formMetadata);
    const { renderInput } = this.inputRenderer;
    const displayInput =  (label) => (
      <div className="size_in">
        <div className="add_dept">
          {renderInput('item', 'filter-dropdown-select',
            {
              className: 'department_dropdown',
              size: '',
              label: label
            })
          }
        </div>
        <button
          type="button"
          onClick={addItem}
          disabled={values.item.length < 1}
          className={values.item.length > 1 ? 'add_button' : 'disable_button'}>
      Add
        </button>
      </div>
    );

    const displayDefaultCenters = () => (
      <div>
        Centers
        <div className="btn-group center-button-group">
          {allCenters.map(center => (
            <button 
              key={Math.floor((Math.random() * 5000) + 1)}
              type="button"
              className="center_buttons">
              {center}
            </button>
          ))}
        </div>
      </div>
    );

    return (
      <fieldset className="personal-details">
        <div>
          <div style={{ paddingTop: '14px' }}>
            {renderInput('email', 'filter-dropdown-select',
              {
                className: 'email_dropdown', size: '', disabled
              })
            }
          </div>

          <div>
            {roleName === 'Budget Checker' ?
              displayInput('Departments') : 
              roleName === 'Super Administrator' && myTitle === 'Add User' ?
                displayDefaultCenters() :
                !(/Manager|Requester/.test(roleName)) && displayInput('Country')
            }
          </div>
          <div className="">
            <table className="mdl-js-data-table table__requests" style={{ backgroundColor: 'white'}}>
              {values.items.map((list, i) =>
                (
                  <tbody key={Math.floor((Math.random() * 5000) + 1)} className="table__body">
                    <tr className="table__row" style={{ backgroundColor: '#F8F8F8' }}>
                      <td className="mdl-data-table__cell--non-numeric table__data" style={{ textTransform: 'capitalize', borderLeft: 'none' }}>
                        {list}
                      </td>
                      <td style={{ borderRight: 'none'}}>
                        <i
                          role="button"
                          className="remove_department"
                          id="remove"
                          onKeyDown={this.handleKeyDown}
                          tabIndex="0"
                          onClick={() =>removeItem(i)}>
                              X
                        </i>
                      </td>
                    </tr>
                  </tbody>
                ))}
            </table>
          </div>
        </div>
      </fieldset>
    );
  }
}

const roleName = PropTypes.string;
const allMails = PropTypes.array;
const values = PropTypes.object;
const addItem = PropTypes.func;
const removeItem = PropTypes.func;
const items = PropTypes.array
;
PersonalDetailsFiedset.propTypes = {
  roleName: roleName.isRequired,
  allMails: allMails.isRequired,
  values: values,
  removeItem: removeItem,
  addItem: addItem,
  myTitle: PropTypes.string,
  departments: PropTypes.array,
  centers: PropTypes.array
};

PersonalDetailsFiedset.defaultProps = {
  removeItem: () => {},
  addItem: () => {},
  values: {},
  myTitle: '',
  departments: [],
  centers: []
};

export default PersonalDetailsFiedset;
