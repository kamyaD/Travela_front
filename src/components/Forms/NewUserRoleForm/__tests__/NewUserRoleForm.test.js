import React from 'react';
import sinon from 'sinon';
import NewUserRoleForm from '../NewUserRoleForm';
import PersonalDetails from '../FormFieldsets/PersonalDetails';

describe('<NewUserRoleForm />', () => {
  let wrapper, onSubmit;
  onSubmit = jest.fn();

  const props = {
    user: {
      UserInfo: {
        name: 'John Doe'
      },
    },
    role: 'travel team member',
    loading: false,
    errors: [],
    myTitle: 'Add User' ,
    getRoleData: jest.fn(() => {}),
    handleUpdateRole:  jest.fn(() => {}),
    onChange: jest.fn(),
    userDetail: {
      email: 'tomato@andela.com',
      id: 1,
      centers: [{
        location: 'New York, USA'
      }]
    },
    centers: [{location: 'Kigali, Rwanda'}],
    roleId: '33589',
  };

  beforeEach(() => {
    wrapper = mount(<NewUserRoleForm {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders two fieldsets', () => {
    const fieldsets = wrapper.find('fieldset');
    expect(fieldsets).toHaveLength(2);
  });

  it('renders the personal details fieldset', () => {
    const personalDetails = wrapper.find('fieldset.personal-details');
    expect(personalDetails).toHaveLength(1);
  });

  it('picks input values', () => {
    wrapper.find('input[name="email"]').simulate('change', {
      target: {
        email: 'email',
        value: 'Tomato Guy'
      }
    });
    expect(wrapper.state().values.email).toBe('Tomato Guy');
  });

  it('validates form before sending data', () => {
    const form = wrapper.find('form');
    form.simulate('submit');
    expect(onSubmit).toHaveBeenCalledTimes(0);
  });

  it('should render a loading indicator while updating a new role', () => {
    wrapper.setProps({ updatingRole: true });
    expect(wrapper.find('h5').text()).toEqual('Updating role...');
  });


  it('calls on submit when all details are correct', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleSubmit');
    wrapper.instance().forceUpdate();
    wrapper.find('form').simulate('submit');
    wrapper.state().values.email = 'test';
    expect(spy).toHaveBeenCalled();
    expect(props.handleUpdateRole).toHaveBeenCalledWith(wrapper.state().values);
    expect(props.handleUpdateRole).toHaveBeenCalledTimes(2);
  });


  it('calls on submit when all details are correct when changing user center ', () => {
    const wrapper =  mount(<NewUserRoleForm {...{...props, myTitle: 'Change Center'}} />);
    const spy = jest.spyOn(wrapper.instance(), 'handleSubmit');
    wrapper.instance().forceUpdate();
    wrapper.find('form').simulate('submit');
    expect(spy).toHaveBeenCalled();
    expect(props.handleUpdateRole).toHaveBeenCalledTimes(2);
  });


  it('should clear the form when the component unmounts', () => {
    const componentWillUnmount = jest.spyOn(wrapper.instance(), 'componentWillUnmount');
    wrapper.unmount();
    expect(componentWillUnmount).toHaveBeenCalledTimes(1);
  });

  it('calls the componentDidMount method', () => {
    const spy = sinon.spy(NewUserRoleForm.prototype, 'componentDidMount');
    const wrapper = mount(
      <NewUserRoleForm {...{...props,  
        role: 'travel team member',  userDetail: {
          email: 'tomato@andela.com',
          id: 1,
          centers: [{
            location: 'New York, USA'
          }]
        }}} />);
    expect(spy.called).toEqual(true);
    wrapper.unmount();
  });

  describe('<PersonalDetails />', () => {
    const props = {
      myTitle: 'Change Center',
      centers: [{
        location: []
      }],
      roleName: 'travel team member',
      validate: true
    };

    beforeEach(() => {
      wrapper = shallow(<PersonalDetails {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

  });

});
