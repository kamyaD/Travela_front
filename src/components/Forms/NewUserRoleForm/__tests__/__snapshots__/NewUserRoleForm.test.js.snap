// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewUserRoleForm /> <PersonalDetails /> renders correctly 1`] = `
<fieldset
  className="personal-details"
>
  <div>
    <div
      style={
        Object {
          "paddingTop": "14px",
        }
      }
    >
      <Input
        autoComplete="off"
        className=""
        disabled={true}
        label="Andela Email Address"
        labelNote=""
        name="email"
        type="text"
      />
    </div>
    <div>
      <Input
        autoComplete="off"
        choices={
          Array [
            Array [],
          ]
        }
        className=""
        label="Center"
        labelNote=""
        name="center"
        type="dropdown-select"
      />
    </div>
    <div />
  </div>
</fieldset>
`;

exports[`<NewUserRoleForm /> renders correctly 1`] = `
<form
  className="new-request"
  onSubmit={[Function]}
>
  <fieldset
    className="personal-details"
  >
    <div>
      <div
        style={
          Object {
            "paddingTop": "14px",
          }
        }
      >
        <div
          className="form-input  "
        >
          <label
            htmlFor="email"
          >
            Andela Email Address
            <span
              style={
                Object {
                  "color": "red",
                }
              }
            >
              *
            </span>
          </label>
          <input
            autoComplete="off"
            className=" "
            disabled={false}
            label="Andela Email Address"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="tomato@andela.com"
          />
          <span
            className="error"
          />
        </div>
      </div>
      <div />
      <div />
    </div>
  </fieldset>
  <hr />
  <fieldset>
    <div
      className="submit-area"
    >
      <button
        className="bg-btn bg-btn--inactive"
        id="cancel"
        onClick={[Function]}
        type="button"
      >
        Clear
      </button>
      <button
        className="bg-btn bg-btn--active"
        disabled={true}
        id="submit"
        type="submit"
      >
        Submit
      </button>
    </div>
  </fieldset>
</form>
`;
